docs/proposals/210701_endpoint_discovery.md

```diff
@@ -0,0 +1,44 @@
+---
+title: Unified Endpoint discovery
+type: proposal
+menu: proposals
+status: XXX
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/2600
+
+### Summary
+
+We want to propose a new flag called `--endpoint=<address>` that will be passed to Thanos query and replace the current `--store=<address>` and `--rule=<address>` flags, and add a discovery mechanism which will be used to determine what types of services this endpoint can serve.
+
+### Motivation
+
+Currently, in Thanos Query, the discovery of rules APIs happens via Store API's Info method, this makes it harder if we ever want to not have a coupling to the Store API (which is already planned for [scalable ruler proposal](https://github.com/thanos-io/thanos/blob/master/docs/proposals/202005_scalable-rule-storage.md)).
+
+We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>` and perform DNS discovery on often the same address multiple times, this can cause occasional DNS issues. Which is especially confusing when one works and the other doesn't.
+
+### Goals
+
+Unite discovery of endpoints, which would make code cleaner as well as provide building blocks to provide discover and support for new gRPC services, for example, the Targets API/Exemplar API/Metadata API.
+
+### Solution
+
+Add a new flag called `--endpoint` to Thanos query, and auto-discover what services that endpoint is serving based on metadata each gRPC server exposes.
+
+The plan would be for the initial discovery of gRPC services available to happen via gRPC reflection. To get further information, we would be adding an Info method to the Rules API. This would contain similar info or metadata information as we get right now from the Store Info method. Potentially in the future, we would implement this same Info method for other services in the, for example, the Targets, Examplers, etc. As mentioned the discovery would happen via reflection initially, the Info method in the Rules API would expose extra metadata about external labels for now, and we can always add more data to it in the future as needed.
+
+The upgrade path would be possible in this case, as it's strictly additive for endpoints, and therefore in theory users would need to just upgrade in the correct order, with Thanos Query component being the last to be upgraded.
+
+#### Alternatives
+
+Split Info method out of the Store API into a separate gRPC service called `info`. Each component must implement and expose this. The discovery would happen via this info gRPC service, which each component would need to explicitly state which API it supports. Besides that information, we would also need to include here various info, e.g. LabelSets, Max/Min time, etc. The drawback for this solution is harder upgrades as this would be more complex to implement as we would need to fall back to older APIs.
```

Comment:
I could see this one working as well if it is a strict subset kind of in the form of:

```yaml
info:
  external_labels: ...
  store:
    minTime: ...
    maxTime: ...
  rule:
    xyz: ...
```

What I do like about this is that we would only have to do a single call to discover all metadata, regardless of the type of store, and through the existence of the sub-section we know that the API is available and ready to serve. (also fine with having a top level field that says whether an API is served is fine by me as well, implicit things tend to go wrong or be forgotten)

Although I do agree, the migration path seems kind of painful code complexity wise. If we should go for this, I would actually suggest to implement this 100% parallel and duplicate, so that when done the "old" code can just be deleted in its entirety.

Reply:
So if I understand correctly we would in this case not go via reflection at all to discover the type, as we would always expect each service to serve this metadata?

Reply:
My bad, I should have been more specific. That is exactly what I meant.

Reply:
@bwplotka What do you think of choosing this option, its similar to your initial idea!

After we agree on it, I will document it in details, does that process sound good? 🙂

Reply:
> @bwplotka What do you think of choosing this option, its similar to your initial idea!

It allows Unified `InfoAPI` so yes!

However still I would say `reflection vs checking fields` is something not agreed. With this structure @brancz  proposed we can avoid reflection, but I still think we can (or maybe have to) use reflection for cases where Info will not include any specific component field.

Reply:
Sounds great to me, thanks for commenting, I think maybe the discovery might make more sense in the actual PR. We can leave it to be not reflection for now here and in case it makes sense to use reflection initially or for some parts there are no objections?

So to sum up:
- we want to use unifed InfoAPI that includes the above metadata
- discovery via that API or reflection depending
- and try to make upgrades work smoothly regardless

@brancz @bwplotka does that sum it up or did I misunderstand any points here? Thank you!

Reply:
That sums it up well, and I agree reflection could come in handy in a migration path, I'd say that would be an implementation detail that can be discussed on the PR though. I think for the scope of this proposal, I'm happy with this.

Reply:
Perfect summary + 💯